contracts/mesh-provider/src/contract.rs

```diff
@@ -253,6 +271,56 @@ pub fn execute_unbond(
     Ok(Response::new().add_message(msg))
 }
 
+// HACK this implementation of claiming rewards is not performant or robust
+// It is intended for proof of concept only.
+pub fn execute_claim_rewards(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    validator: String,
+) -> Result<Response, ContractError> {
+    let delegator = deps.api.addr_validate(info.sender.as_str())?;
+    let config = CONFIG.load(deps.storage)?;
+
+    // calculate rewards
+    let validator_info = VALIDATORS.load(deps.storage, &validator)?;
+    let mut delegator_stake = STAKED.load(deps.storage, (&delegator, &validator))?;
+
+    // We calculate the rewards
+    delegator_stake.calc_pending_rewards(
+        validator_info.rewards.total_rptpb,
+        delegator_stake.shares,
+        env.block.height,
+    )?;
+
+    if delegator_stake.rewards.pending.is_zero() {
+        return Err(ContractError::NoRewardsToClaim {});
+    }
+
+    let balance = deps
+        .querier
+        .query_balance(env.contract.address, config.denom.clone())?;
+
+    // Make sure we have something to send, if its false, funds might be stuck in consumer and need admin. (or we messed up badly)
+    if delegator_stake.rewards.pending > balance.amount {
+        return Err(ContractError::WrongBalance {
+            balance: balance.amount,
+            rewards: delegator_stake.rewards.pending,
+        });
+    }
+
+    let msg = BankMsg::Send {
+        to_address: delegator.to_string(),
+        amount: vec![coin(delegator_stake.rewards.pending.u128(), config.denom)],
```

Comment:
This is a bit confusing to read but totally correct.

maybe `denom` -> `reward_denom` or `ibc_denom`?
to clarify it is not local denom being staked.